### PR TITLE
GMP: Make results ternary and don't fail on callee revert

### LINKED ISSUE #623: Catch revert in cross network message execution
Catch revert in cross network message execution in `LibGateway` so that bottom up checkpoint submission will not be blocked.

### LINKED ISSUE #624: Add type information to ReceiptMsg returned bytes
Currently we are mixing cross message returned bytes, there is no way for the user to know if the returned bytes are coming from IPC execution or target contract reversion. One can make a guess by inspecting the returned bytes length, but  we can make the explicit by specifying the type so that we dont need to guess.

### PR DESCRIPTION
**Pull request by cryptoAtwill**
This PR contains two fixes to existing checkpoint execution:

1. Adding an enum `ReceiptType` to `ReceiptMsg`, so that when a receipt is returned, the user knows how to parse the `ret` bytes, whether it's an IPC error selector or contract call revert bytes, or other potential uses. This should fix #624. 

2. Currently the execution of cross messages does not catch revert. If the execution reverts, then the checkpoint commitment is also reverted. This means the submission is blocked. We now will catch the revert and return as part of the receipt. This fixes #623.

### CODE REVIEW

File: contracts/src/lib/LibGateway.sol [hunk 1]
```diff
@@ -412,8 +412,22 @@ library LibGateway {
...
 
         // execute the message and get the receipt.
-        (bool success, bytes memory ret) = crossMsg.execute(supplySource);
-        sendReceipt(crossMsg, success, ret);
+        (bool success, bytes memory ret) = executeCrossMsg(crossMsg, supplySource);
+        sendReceipt(crossMsg, ReceiptType.InvokedContract, success, ret);
+    }
+
+    /// @dev Execute the cross message using low level `call` method. This way ipc will
+    /// @dev catch contract revert messages as well
```
Comment (dnkolegov) [inline contracts/src/lib/LibGateway.sol:420]: @dev typo

**Review by dnkolegov — APPROVED** [review]

**Review by dnkolegov — COMMENTED** [review]
It would be great to fix NATSPEC descriptions. All @dev are duplicates

File: contracts/src/lib/CrossMsgHelper.sol [hunk 2]
```diff
@@ -71,10 +71,11 @@ library CrossMsgHelper {
     /// and identifies the receipt through the hash of the original message.
     function createReceiptMsg(
         IpcEnvelope calldata crossMsg,
+        ReceiptType receiptType,
         bool success,
         bytes memory ret
     ) public pure returns (IpcEnvelope memory) {
-        ReceiptMsg memory message = ReceiptMsg({id: toHash(crossMsg), success: success, ret: ret});
+        ReceiptMsg memory message = ReceiptMsg({receiptType: receiptType, id: toHash(crossMsg), success: success, ret: ret});
```
Comment (raulk) [inline contracts/src/lib/CrossMsgHelper.sol:78]: @cryptoAtwill This is a bit entropic, let's simplify:

- remove `receiptType`
- rename `success` to `outcome`
- add an `outcomeType` enum with three values: `Ok`, `ActorErr`, `SystemErr`.

**Review by raulk — CHANGES_REQUESTED** [review]
I just merged a PR that conflicts with this one, mostly because it renames ReceiptMsg to ResultMsg to avoid ambiguity. The rationale is that a _receipt_ tends to refer to the output of a transaction and belongs in the chain schema. We already have receipts elsewhere, so calling this receipt is confusing.

I'll take care of resolving the conflicts here, and you can then go ahead and apply the change I proposed in my review.

EDIT: conflicts resolved.

File: contracts/src/structs/CrossNet.sol [hunk 3]
```diff
@@ -93,11 +93,23 @@ struct CallMsg {
     bytes params;
 }
 
+/// @notice This struct indicates if the cross message execution is sucess, IPC system error or from the invoked
+///         contract
+enum OutcomeType {
+    /// @dev The execution is successful, parse the return bytes according to the contract logic
+    Ok,
+    /// @dev The receipt is from IPC contract, parse the return bytes as error selector
```
Comment (raulk) [inline contracts/src/structs/CrossNet.sol:101]:
```suggestion
    /// @dev The result is an IPC system error, parse the return bytes as an IPC error type.
```

**Comment by cryptoAtwill** [comment]
@raulk Nope, this PR won't do that. I'm working on the Transfer receipt, will create another PR on top of this.

File: contracts/src/lib/LibGateway.sol [hunk 4]
```diff
@@ -412,8 +412,26 @@ library LibGateway {
...
+            sendReceipt(crossMsg, OutcomeType.Ok, ret);
+        } else {
+            sendReceipt(crossMsg, OutcomeType.ActorErr, ret);
+        }
+    }
+
+    /// @dev Execute the cross message using low level `call` method. This way ipc will
+    /// @dev catch contract revert messages as well
+    function executeCrossMsg(IpcEnvelope memory crossMsg, SupplySource memory supplySource) internal returns (bool success, bytes memory result) {
+        (success, result) = address(CrossMsgHelper).delegatecall(   // solhint-disable-line avoid-low-level-calls
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply contracts/src/lib/LibGateway.sol:428]: yep, added.

**Pull request merged**